crypto/openssl_*: Standardize implementaton and add support keyUsage, extenededKeyUsage

cpointner opened the pull request:
##### SUMMARY

The OpenSSLObject class has been merged[1]. This commit makes the
openssl_csr module rely on this class and standardize the way openssl
module should be written.
This also adds checks for subject alt names as well as support for keyUsage
and extendedKeyUsage certificate extensions.

This is based on the work of @psykotox and obsoletes #25215.

[1] #26945

##### ISSUE TYPE

 - Feature Pull Request

##### COMPONENT NAME

  - openssl_csr

##### ANSIBLE VERSION
```
devel
```

##### ADDITIONAL INFORMATION
```
N/A
```

Commit:
openssl_csr: make subjectAltNames a list

Commit:
csr module now uses the new standard way to build openssl crypto modules

Commit:
add check functions for subject and subjectAltNames

Commit:
added support for keyUsage and extendedKeyUsage

Commit:
check if CSR signature is correct (aka the privatekey belongs to the CSR)

Spredzy commented on lib/ansible/modules/crypto/openssl_csr.py:
```diff
@@ -109,6 +105,28 @@
+            - This can either be a 'comma separated string' or a YAML list.
+
+notes:
+    - "If the certificate signing request already exists it will be checked whether subjectAltName,
```
Although to mention that it will `verify()` the privatekey it has been generated with.

Spredzy commented on lib/ansible/modules/crypto/openssl_csr.py:
```diff
@@ -165,40 +194,54 @@
+        self.subjectAltName = module.params['subjectAltName']
+        self.keyUsage = module.params['keyUsage']
+        self.extendedKeyUsage = module.params['extendedKeyUsage']
         self.changed = True
```
This isn't needed anymore as it is part of the parent class value. **Note**: Default has been set to `False`, so the logique about self.changed needs to be updated in the rest of the code.

Spredzy commented on lib/ansible/modules/crypto/openssl_csr.py:
```diff
@@ -254,16 +302,79 @@ def generate(self, module):
-        except OSError as exc:
-            if exc.errno != errno.ENOENT:
-                raise CertificateSigningRequestError(exc)
+        self.privatekey = crypto_utils.load_privatekey(self.privatekey_path, self.privatekey_passphrase)
```
Not sure I understand the `self.privatekey`` here.

I'd say it belongs circa l. 367 within the `def _check_signature()` method as this is the only place where it is need.

cpointner replied:
No it is also needed when signing the certificate request (see L289) and i think it is ugly to open the private key twice. Opening it in __init__ is of course a very bad idea so this is actually the only place where i can be sure that the key gets opened before an existing CSR gets checked or a new one is signed.

Spredzy replied:
Fair engough. Then, if I haven't overlooked anything, there is no need to pass `privatekey` to `_check_signature()` as you can call `csr.verify(self.privatekey)` directly.

cpointner replied:
No i can't because `_check_signature` as all the other _check functions is no member function of the class

cpointner replied:
...which is wrong of course, because the _check functions are closures. Stupid me... Sorry for the noise!

Spredzy commented on lib/ansible/modules/crypto/openssl_csr.py:
```diff
@@ -272,6 +383,8 @@ def dump(self):
             'csr': self.path,
```
nit: can you:

 * rename `csr` to `filename` to matches privatekey and publickey module. So as the user, one would always know what is the return parameter s/he needs to look at.
 * add 'privatekey': self.privatekey_path, so a user would know which privatekey has been used to generate that csr

Spredzy reviewed (changes requested):
Thanks @cpointner for taking up on this work. First review is done, minor changes required. I'll just dig deeper on the KeyUsage and extendedKeyUsage on my side, just to ensure I get everything you're doing on the check.

In general, LGTM on this review after comments are addressed

Commit:
fixes for first PR review

Spredzy commented on lib/ansible/modules/crypto/openssl_csr.py:
```diff
@@ -247,31 +295,97 @@ def generate(self, module):
+
+            return True
+
+        def _check_keyUsage_(extensions, extName, expected, long):
```
This check doesn't work as-is and raises issues when running the following playbook two times in a row:

```
---
- hosts: localhost
  tasks:
    - openssl_privatekey:
        path: /tmp/private.key

    - openssl_csr:
        path: /tmp/csr.csr
        privatekey_path: /tmp/private.key                                                                                                                                                      
        commonName: www.ansible.com
```
This is due to the fact thatthe current code relies on `expected` being an array when it could be actually None, hence raising:

```
TypeError: 'NoneType' object is not iterable
```

Here is an alternative implentation that did what was expected, feel free to modify adapt/modify/get ideas from it:

```
usages_ext = [str(ext) for ext in extensions if ext.get_short_name() == extName]

if (not usages_ext and expected) or (usages_ext and not expected):
    return False
elif not usages_ext and not expected:
    return True
else:
    current = [usage.strip() for usage in usages_ext[0].split(',')]
    expected = [long[usage] for usage in expected]
    return current == expected
```

cpointner replied:
I more or less used your proposed check function. It is indeed much nicer than my version.
I only had to slightly change the list comprehension for `expected`. The reason for this is that the user supplied value in `expected` might already be the long version of the usage name. Openssl accepts both the long and the short version when creating the CSR but will always returns the long version. For this reason i need to look up the long name for usages where we got the short name from the user.

Commit:
fixes for second PR review

cpointner commented:
Closing & reopening to trigger CI

Commit:
openssl_csr: there is no need to pass on privatekey as it can be accessed directly

cpointner commented:
@Spredzy thanks for your review. I think i have addressed all of your comments. If you prefer i can squash my changes into one or two commits to make it easier to review the changes.

Spredzy commented on lib/ansible/modules/crypto/openssl_csr.py:
```diff
@@ -296,6 +398,9 @@ def main():
             organizationalUnitName=dict(aliases=['OU'], type='str'),
             commonName=dict(aliases=['CN'], type='str'),
             emailAddress=dict(aliases=['E'], type='str'),
+            subjectAltName=dict(aliases=['subjectAltName'], type='list'),
```
Not your fault, but can you remove the aliases[] please, since its the same value

Spredzy commented on lib/ansible/modules/crypto/openssl_csr.py:
```diff
@@ -247,31 +295,86 @@ def generate(self, module):
 
         result = {
-            'csr': self.path,
+            'privatekey': self.privatekey_path,
```
Can you updated the RETURN documentation to:

  * s/csr/filename
  * add `keyUsage` and `extendedUsage`

Spredzy reviewed (changes requested):
Two minor things and then shipit!

For the commit squash do not worry about it, I think that the bot or the merge strategy will eventually squash them all in one.

Commit:
openssl_csr: documentation fixes

cpointner commented:
bot_status

cpointner commented:
ready_for_review

Spredzy commented:
shipit

Spredzy commented:
bot_status

cpointner commented:
@Spredzy: i'm a little confused... the bot says it's waiting for me but i don't know what needs to be done here. Do you know whats missing?

Spredzy commented:
bot_status

Spredzy commented:
@cpointner The waiting_on: maintainer should be solved once https://github.com/ansible/ansible/pull/27691 is merged.

This is because `modules_utils/crypto.py` is modified and I am not listed as the maintainer, hence the message. Will redo a bot_status once merged

Spredzy commented:
bot_status

gundalow commented:
bot_status

Spredzy commented:
Thanks @cpointner for this work. Much appreciated :)
